modules/auxiliary/scanner/rsync/modules_list.rb

```diff
@@ -24,46 +23,97 @@ def initialize
     )
     register_options(
       [
-        Opt::RPORT(873)
+        Opt::RPORT(873),
+        OptBool.new('AUTH_CHECK', [true, 'Check authentication or not', false])
       ], self.class)
   end
 
-  def run_host(ip)
+  def rsync(dir)
     connect
-    version = sock.get_once
 
+    version = sock.get_once # server_initialisation
     return if version.blank?
 
-    print_good("#{ip}:#{rport} - rsync #{version.strip} found")
-    report_service(:host => ip, :port => rport, :proto => 'tcp', :name => 'rsync')
-    report_note(
-        :host => ip,
-        :proto => 'tcp',
-        :port => rport,
-        :type => 'rsync_version',
-        :data => version.strip
-    )
-
-    # making sure we match the version of the server
-    sock.puts("#{version}")
-    # the listing command
-    sock.puts("\n")
-    listing = sock.get(20)
+    sock.get(3) # server_motd
+    sock.puts(version) # client_initialisation
+    sock.puts(dir) # client_query
+    data = sock.get(3) # module_list
+    data.gsub!('@RSYNCD: EXIT', '')
     disconnect
+    [version, data]
+  end
+
+  def auth?(dir)
+    _version, data = rsync(dir)
+    if data && data =~ /RSYNCD: OK/m
+      vprint_status("#{dir} needs authentication: false")
+      false
+    else
+      vprint_status("#{dir} needs authentication: true")
+      true
```

Comment:
This feels incorrect to me, perhaps more-so given my previous comment about the lack of rsync protocol verification when you first connect.  This will incorrectly report that the rsync module supports authentication when:

* `data` is falsey
* `data` does not include the string "RSYNCD: OK" somewhere in its response.   It could just include a pile of random characters as a response, which is not authentication.

IMO, you should explicitly check for `@RSYNCD: AUTHREQD`, and only then should you return true.